### ISSUE TITLE
False positive `unnecessary-lambda` when the lambda refers to a not-yet-defined symbol

### ISSUE THREAD

**Issue by jace** [pylint-dev/pylint#9306]
### Bug description

```python
class Foo:
    def __init__(self, a): self.a = a
    def __call__(self): return self.a()

f = Foo(lambda: future_func())

def future_func():
    return "future"

print(f())
```

### Configuration

```ini
[MESSAGE CONTROL]
enable=
    unnecessary-lambda
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a.py
a.py:5:8: W0108: Lambda may not be necessary (unnecessary-lambda)
```

### Expected behavior

No error. The lambda is necessary because it's referring to a symbol that has not yet been defined. Pylint should not flag the lambda in such a case.

This was spotted when defining a SQLAlchemy relationship's `order_by` parameter, [where the documentation explicitly recommends a lambda to resolve future symbols](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.order_by). In my case, this looked like:
```python
from __future__ import annotations
from sqlalchemy.orm import Mapped, relationship

class ParentModel:
    children: Mapped[list[ChildModel]] = relationship(order_by=lambda: ChildModel.date.desc())

class ChildModel:
    ...
```

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Oct  2 2023, 13:45:54) [Clang 15.0.0 (clang-1500.0.40.1)]
```
